v2/codetags/parser.go

```diff
@@ -0,0 +1,372 @@
+			case unicode.IsSpace(r):
+				continue
+			case r == ')':
+				incomplete = false
+				st = stMaybeValue
+			case r == '0':
+				buf.WriteRune(r)
+				st = stPrefixedNumber
+			case r == '-' || r == '+' || unicode.IsDigit(r):
+				buf.WriteRune(r)
+				st = stNumber
+			case r == '"' || r == '`':
+				quote = r
+				st = stQuotedString
+			case isIdentBegin(r):
+				buf.WriteRune(r)
+				st = stNakedString
+			default:
+				break parseLoop
+			}
+		case stNumber:
+			hexits := "abcdefABCDEF"
+			switch {
+			case unicode.IsDigit(r) || strings.Contains(hexits, string(r)):
+				buf.WriteRune(r)
+				continue
+			case r == ',':
+				if err := saveInt(); err != nil {
+					return tagKey{}, err
+				}
+				st = stArg
+			case r == ')':
+				if err := saveInt(); err != nil {
+					return tagKey{}, err
+				}
+				incomplete = false
+				st = stMaybeValue
+			case unicode.IsSpace(r):
+				if err := saveInt(); err != nil {
+					return tagKey{}, err
+				}
+				st = stEndOfToken
+			default:
+				break parseLoop
+			}
+		case stPrefixedNumber:
+			switch {
+			case unicode.IsDigit(r):
+				buf.WriteRune(r)
+				st = stNumber
+			case r == 'x' || r == 'o' || r == 'b':
+				buf.WriteRune(r)
+				st = stNumber
+			default:
+				break parseLoop
+			}
+		case stQuotedString:
+			switch {
+			case r == '\\':
+				st = stEscape
+			case r == quote:
+				saveString()
+				st = stEndOfToken
+			default:
+				buf.WriteRune(r)
+			}
+		case stEscape:
+			switch {
+			case r == quote || r == '\\':
+				buf.WriteRune(r)
+				st = stQuotedString
+			default:
+				return tagKey{}, fmt.Errorf("unhandled escaped character %q", r)
+			}
+		case stNakedString:
+			switch {
+			case isIdentInterior(r):
+				buf.WriteRune(r)
+			case r == ',':
+				saveBoolOrString()
+				st = stArg
+			case r == ')':
+				saveBoolOrString()
+				incomplete = false
+				st = stMaybeValue
+			case unicode.IsSpace(r):
+				saveBoolOrString()
+				st = stEndOfToken
+			case r == ':':
+				saveName()
+				st = stArg
+			default:
+				break parseLoop
+			}
+		case stEndOfToken:
+			switch {
+			case unicode.IsSpace(r):
+				continue
+			case r == ',':
+				st = stArg
+			case r == ')':
+				incomplete = false
+				st = stMaybeValue
+			default:
+				break parseLoop
+			}
+		case stMaybeValue:
+			switch {
+			case r == '=':
+				st = stValue
+			case unicode.IsSpace(r):
+				st = stMaybeComment
+			default:
+				break parseLoop
+			}
+		case stValue: // This is a terminal state, it consumes the rest of the input as an opaque value.
+			value.WriteRune(r)
+		case stMaybeComment:
+			switch {
+			case unicode.IsSpace(r):
+				continue
+			case r == '/':
+				incomplete = true
+				st = stTrailingSlash
+			default:
+				break parseLoop
+			}
+		case stTrailingSlash:
+			switch {
+			case r == '/':
+				incomplete = false
+				st = stTrailingComment
+			default:
+				break parseLoop
+			}
+		case stTrailingComment:
+			i = len(runes) - 1
+			break parseLoop
+		default:
+			panic("unknown state")
```

Comment:
I know it's kind an `ASSERT()` but it still feels dirty to panic.

And anyway, you could add the string of the unhandled state :)